shell/platform/android/android_image_generator.cc

```diff
@@ -0,0 +1,192 @@
+  if ((status = AndroidBitmap_lockPixels(env, bitmap->obj(), &pixel_lock)) <
+      0) {
+    FML_DLOG(ERROR) << "Failed to lock pixels, error=" << status;
+    return;
+  }
+
+  SkData::ReleaseProc on_release = [](const void* ptr, void* context) -> void {
+    fml::jni::ScopedJavaGlobalRef<jobject>* bitmap =
+        reinterpret_cast<fml::jni::ScopedJavaGlobalRef<jobject>*>(context);
+    auto env = fml::jni::AttachCurrentThread();
+    AndroidBitmap_unlockPixels(env, bitmap->obj());
+  };
+
+  software_decoded_data_ = SkData::MakeWithProc(
+      pixel_lock, info.width * info.height * sizeof(uint32_t), on_release,
+      bitmap);
+}
+
+bool AndroidImageGenerator::Register(JNIEnv* env) {
+  g_flutter_jni_class = new fml::jni::ScopedJavaGlobalRef<jclass>(
+      env, env->FindClass("io/flutter/embedding/engine/FlutterJNI"));
+  FML_DCHECK(!g_flutter_jni_class->is_null());
+
+  g_decode_image_method = env->GetStaticMethodID(
+      g_flutter_jni_class->obj(), "decodeImage",
+      "(Ljava/nio/ByteBuffer;J)Landroid/graphics/Bitmap;");
+  FML_DCHECK(g_decode_image_method);
+
+  static const JNINativeMethod header_decoded_method = {
+      .name = "nativeImageHeaderCallback",
+      .signature = "(JII)V",
+      .fnPtr = reinterpret_cast<void*>(
+          &AndroidImageGenerator::NativeImageHeaderCallback),
+  };
+  if (env->RegisterNatives(g_flutter_jni_class->obj(), &header_decoded_method,
+                           1) != 0) {
+    FML_LOG(ERROR)
+        << "Failed to register FlutterJNI.nativeImageHeaderCallback method";
+    return false;
+  }
+
+  return true;
+}
+
+std::unique_ptr<ImageGenerator> AndroidImageGenerator::MakeFromData(
+    sk_sp<SkData> data,
+    fml::RefPtr<fml::TaskRunner> task_runner) {
+  auto* generator = new AndroidImageGenerator(std::move(data));
+
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runner, [generator]() { generator->DecodeImage(); });
+
+  if (generator->IsValidImageData()) {
+    return std::unique_ptr<AndroidImageGenerator>(generator);
+  }
+  return nullptr;
+}
+
+void AndroidImageGenerator::NativeImageHeaderCallback(JNIEnv* env,
+                                                      jclass jcaller,
+                                                      jlong generator_address,
+                                                      int width,
+                                                      int height) {
+  AndroidImageGenerator* generator =
+      reinterpret_cast<AndroidImageGenerator*>(generator_address);
+
+  generator->image_info_ = SkImageInfo::Make(
+      width, height, kRGBA_8888_SkColorType, kPremul_SkAlphaType);
+  generator->header_decoded_latch_.Signal();
+}
+
+bool AndroidImageGenerator::IsValidImageData() {
+  // The generator kicks off an IO task to decode everything, and calls to
+  // "GetInfo()" block until the header has been decoded. The decoder explicitly
+  // marks the height as -1 if the image is invalid or if the SDK image decoder
+  // is unavailable.
```

Comment:
Is this true? I'm not seeing any IO tasks getting kicked off, decoding is all happening on the platform thread right?

Reply:
`AndroidImageGenerator::MakeFromData` kicks off the task and is the only reasonable way to produce this `ImageGenerator`.